pkg/controller/configstatus/configstatus_controller.go

```diff
@@ -0,0 +1,195 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configstatus
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/go-logr/logr"
+	configv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/apis/status/v1beta1"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/util"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller").WithValues(logging.Process, "config_status_controller")
+
+type Adder struct {
+	WatchManager *watch.Manager
+}
+
+func (a *Adder) InjectControllerSwitch(_ *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(_ *readiness.Tracker) {}
+
+// Add creates a new config Status Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (a *Adder) Add(mgr manager.Manager) error {
+	r := newReconciler(mgr)
+	return add(mgr, r)
+}
+
+// newReconciler returns a new reconcile.Reconciler.
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileConfigStatus{
+		// Separate reader and writer because manager's default client bypasses the cache for unstructured resources.
+		writer:       mgr.GetClient(),
+		statusClient: mgr.GetClient(),
+		reader:       mgr.GetCache(),
+
+		scheme: mgr.GetScheme(),
+		log:    log,
+	}
+}
+
+// PodStatusToConfigMapper correlates a ConfigPodStatus with its corresponding Config.
+// `selfOnly` tells the mapper to only map statuses corresponding to the current pod.
+func PodStatusToConfigMapper(selfOnly bool) handler.TypedMapFunc[*v1beta1.ConfigPodStatus] {
```

Comment:
It looks like `selfOnly` is not being used. This code was copied from expansion templates, which also did not use `selfOnly`. This is a bug. Here is an example of proper use of `selfOnly` in the constraint controller:

https://github.com/open-policy-agent/gatekeeper/blob/07127f2352e9793a8471bfd9d1e15e97d69bf81b/pkg/controller/constraint/constraint_controller.go#L194-L198

Here is the basic reasoning:

* status controllers want to watch all podStatus objects and the primary object because they want to make sure they respond to any changes (i.e. writing status changes, overwriting inappropriate deletes of the status field, etc.)

* primary controllers want to watch podStatus for the corresponding pod -- if someone deletes a podStatus resource the main object should be re-reconciled to avoid missing state.

As it stands, if a user were to delete the podStatus object, there is a risk that there would be no reconcile.

We should add the appropriate watches to the config and expansion controllers.

Reply:
Updated.

Reply:
Could we have a PR that fixes this in the other controller?

Reply:
If selfOnly is always required and true, why is this a variable? when does it need to be set to false?